### PR TITLE
integration of Chocolate Doom network code

### LINKED ISSUE #279: feature request: add -solo-net parameter
Received via PM on DW.

### PR DESCRIPTION
**Pull request by rfomin**
I think this is a much cleaner attempt. Original code changes are kept at a minimum, take a look if you have time. Connects successfully to Crispy Doom with `complevel vanilla`. Also fixes #279.

### CODE REVIEW

File: Source/d_loop.c [hunk 1]
```diff
@@ -0,0 +1,864 @@
...
+
+boolean D_InitNetGame(net_connect_data_t *connect_data)
+{
+    boolean result = false;
+    net_addr_t *addr = NULL;
+    int i;
+
+    // Call D_QuitNetGame on exit:
+
+    //I_AtExit(D_QuitNetGame, true);
```
Comment (rfomin) [inline Source/d_loop.c:489]: It seems that the server cannot be properly closed with the standard `atexit()`. Should we implement `I_AtExit()`?

Comment (fabiangreffrath) [reply Source/d_loop.c:489]: > Should we implement I_AtExit()?

We probably should (using the implementation in Pr+). But that's orthogonal to this PR. Let's not mix things up too much.

**Comment by rfomin** [comment]
I would like this to be reviewed.
Known Issues:
1. fastdemo doesn't work.
2. No netgame UI (by default, it just waits for 2 players to connect).
3. Interpolation doesn't work properly in newsync mode.
4. Coop statistics are different from PrBoom+.

**Comment by fabiangreffrath** [comment]
> 1. fastdemo doesn't work.

This only applies to network demos? For singleplayer demos it still works, right?

> 2. No netgame UI (by default, it just waits for 2 players to connect).

This is expected, as we only add the minimal possible amount of libtextscreen.

> 3. Interpolation doesn't work properly in newsync mode.

This is inherited from Crispy, I guess.

https://github.com/fabiangreffrath/crispy-doom/issues/747

> 4. Coop statistics are different from PrBoom+.

This is probably the same as here:

https://github.com/fabiangreffrath/crispy-doom/issues/778

File: Source/d_main.c [hunk 2]
```diff
@@ -1983,11 +1986,15 @@ void D_DoomMain(void)
   if ((p = M_CheckParm ("-timer")) && p < myargc-1 && deathmatch)
     {
       int time = atoi(myargv[p+1]);
+      timelimit = time;
```
Comment (fabiangreffrath) [inline Source/d_main.c:1989]: This already has an implementation in `p_spec.c:P_SpawnSpecials()`. Let's please use one or the other.

File: Source/d_main.c [hunk 3]
```diff
@@ -2230,7 +2243,7 @@ void D_DoomMain(void)
         {
           I_StartTic ();
           D_ProcessEvents ();
-          G_BuildTiccmd (&netcmds[consoleplayer][maketic%BACKUPTICS]);
```
Comment (fabiangreffrath) [inline Source/d_main.c:2233]: What is the `[maketic%BACKUPTICS]` part and why was it dropped?

File: Source/d_net.c [hunk 4]
```diff
@@ -1,580 +1,270 @@
-// Emacs style mode select   -*- C++ -*-
```
Comment (fabiangreffrath) [inline Source/d_net.c:1]: Is this the pristine `d_net.c` file from Choco?

File: Source/CMakeLists.txt [hunk 5]
```diff
@@ -49,6 +49,18 @@ set(WOOF_SOURCES
     m_swap.h
     midifile.c midifile.h
     mmus2mid.c mmus2mid.h
+    net_client.c net_client.h
```
Comment (fabiangreffrath) [inline Source/CMakeLists.txt:52]: Are all the other `net_*` files taken pristine from Choco or are there any changes that need to get documented?

File: Source/CMakeLists.txt [hunk 6]
```diff
@@ -82,6 +94,7 @@ set(WOOF_SOURCES
     r_things.c r_things.h
     s_sound.c  s_sound.h
     s_musinfo.c s_musinfo.h
+    sha1.c     sha1.h
```
Comment (fabiangreffrath) [inline Source/CMakeLists.txt:97]: What do wen need this for? Do we even need this at all?

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]

File: Source/d_net.c (hunk 4 above)
Comment (rfomin) [reply Source/d_net.c:1]: Yes, but without the support of other games (heretic/hexen/strife)

File: Source/d_main.c (hunk 3 above)
Comment (rfomin) [reply Source/d_main.c:2233]: netcmds declaration is different in MBF, I replace it with Choco version.

File: Source/CMakeLists.txt (hunk 5 above)
Comment (rfomin) [reply Source/CMakeLists.txt:52]: The only changes are for other games and for fixing gcc/clang warnings.

File: Source/CMakeLists.txt (hunk 6 above)
Comment (rfomin) [reply Source/CMakeLists.txt:97]: Currently, we don't use it. I guess I should remove it.

**Comment by rfomin** [comment]
> > 1. fastdemo doesn't work.
> 
> This only applies to network demos? For singleplayer demos it still works, right?

It doesn't  work in all types of demos due to the `GetAdjustedTime()` function.

File: Source/d_main.c (hunk 2 above)
Comment (rfomin) [reply Source/d_main.c:1989]: Hmm, I wonder why this is in `P_SpawnSpecials()`?

File: Source/CMakeLists.txt (hunk 5 above)
Comment (rfomin) [reply Source/CMakeLists.txt:54]: Also, I commented out "SecureDemo" experiment.

File: Source/d_net.c (hunk 4 above)
Comment (rfomin) [reply Source/d_net.c:1]: Also, I added sync of Boom/MBF options.

File: Source/net_client.c [hunk 7]
```diff
@@ -0,0 +1,1228 @@
...
+}
+
+void NET_CL_Init(void)
+{
+    // Try to set from the USER and USERNAME environment variables
+    // Otherwise, fallback to "Player"
+
+    if (net_player_name == NULL)
+    {
+        net_player_name = NET_GetRandomPetName();
```
Comment (fabiangreffrath) [inline Source/net_client.c:1214]: For some odd reason, I have my problems with this "pet name" thing. I guess it's too much of an Choco inside joke to apply here. Could you replace this initialization with a `GetUserName()` on WIN32 and `getlogin()` elsewhere?

File: Source/CMakeLists.txt (hunk 5 above)
Comment (fabiangreffrath) [reply Source/CMakeLists.txt:54]: Never heard of that. 🤷

File: Source/d_net.c (hunk 4 above)
Comment (fabiangreffrath) [reply Source/d_net.c:1]: I see this is also where `lowres_turn` hides...

File: Source/CMakeLists.txt (hunk 5 above)
Comment (rfomin) [reply Source/CMakeLists.txt:54]: > 

Look at `NET_StartSecureDemo` in `net_query.c` Not sure how it works.

File: Source/net_client.c (hunk 7 above)
Comment (rfomin) [reply Source/net_client.c:1214]: Maybe it's better to name them "Player 1" ... "Player 4"?

Comment (fabiangreffrath) [reply Source/net_client.c:1214]: Na, that's a bit lame, isn't it? 😉 I think even Vanilla had the sprite colors as player names, i.e. "green", "brown", "indigo", "red".

File: Source/CMakeLists.txt (hunk 5 above)
Comment (fabiangreffrath) [reply Source/CMakeLists.txt:54]: It doesn't seem to be actually used anywhere.

File: Source/net_client.c (hunk 7 above)
Comment (rfomin) [reply Source/net_client.c:1214]: Maybe we should invent dogs related joke here :smile:

Comment (fabiangreffrath) [reply Source/net_client.c:1214]: 😮

Comment (fabiangreffrath) [reply Source/net_client.c:1214]: https://www.theodysseyonline.com/cliche-pet-names

Comment (fabiangreffrath) [reply Source/net_client.c:1214]: Even better:
https://www.dailypaws.com/dogs-puppies/dog-names/paw-patrol-dog-names

Comment (rfomin) [reply Source/net_client.c:1214]: Haha, that's nice.

Comment (rfomin) [reply Source/net_client.c:1214]: net

**Comment by fabiangreffrath** [comment]
Merge it?

**Comment by rfomin** [comment]
> Merge it?

Yes, I think we can. Everything works except fastdemo.

**Comment by fabiangreffrath** [comment]
Yeeha!

Fastdemo will require a separate fix, but it was more important to finally get this networking code merged before.

**Pull request merged**